subcommands/keys/ca_create.go

```diff
@@ -0,0 +1,137 @@
+package keys
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+
+	"github.com/foundriesio/fioctl/subcommands"
+)
+
+var (
+	createOnlineCA bool
+	createLocalCA  bool
+	hsmModule      string
+	hsmPin         string
+	hsmSoPin       string
+)
+
+func init() {
+	cmd := &cobra.Command{
+		Use:   "create <PKI Directory>",
+		Short: "Create PKI infrastructure to manage mutual TLS for the device gateway",
+		Run:   doCreateCA,
+		Args:  cobra.ExactArgs(1),
+		Long: `Peform a one-time operation to set up PKI infrastructure for managing
+the device gateway. This command creates a few things:
+
+# Root of trust for your factory: factory_ca.key / factory_ca.pem
+The factory_ca keypair is generated by this command to define the PKI root of
+trust for this factory.
+ * factory_ca.key - An EC prime256v1 private key that should be STORED OFFLINE.
+ * factory_ca.pem - The public x509 certificate that is shared with
+                    Foundries.io. Once set, all future PKI related changes will
+                    require proof you own this certificate.
+
+# online-ca - A Foundries.io owned keypair to support lmp-device-register
+In order for lmp-device-register to work, Foundries.io needs the ability to
+create client certificates for devices. If enabled, the factory_ca keypair will
+sign the certificate signing request returned from the API.
+
+This is optional.
+
+# local-ca - A keypair you own
+This keypair can be used for things like your manufacturing process where you
+may set up devices without having to communicate with Foundries.io web
+services. This keypair is capable of create client certificates for devices.
+If enabled, the local-ca.pem will be shared with the Foundries.io device gateway
+so that it will trust the client certificate of devices created with this
+keypair.
+
+This is optional.`,
+	}
+	caCmd.AddCommand(cmd)
+	cmd.Flags().BoolVarP(&createOnlineCA, "online-ca", "", true, "Create an online CA owned by Foundries that works with lmp-device-register")
+	cmd.Flags().BoolVarP(&createLocalCA, "local-ca", "", true, "Create a local CA that you can use for creating your own device certificates")
+	cmd.Flags().StringVarP(&hsmModule, "hsm-module", "", "", "Create key on an HSM using this module.")
+	cmd.Flags().StringVarP(&hsmPin, "hsm-pin", "", "", "The PKCS#11 PIN to set up on the HSM, if using one.")
+	cmd.Flags().StringVarP(&hsmSoPin, "hsm-so-pin", "", "", "The PKCS#11 Security Officer PIN to set up on the HSM, if using one.")
```

Comment:
Here I believe we should really just as for the pin, and probably the token ID/name, because most of the times the token is already initialized by the user (and possibly containing extra keys). The init process will erase all other keys if previously created, so only good if we're assuming the user is bootstraping a new HSM.

I believe it is best to assume the user will initialize the token outside this command.

Reply:
i was copying lmp-device-register, but I agree it makes things a pain.